File: webbugs-common/src/models/page.ts
import { Cell, CellType } from "./cell";
import { Coordinates } from "./coordinates";
// @ts-ignore
import _ from 'lodash';

// Page items are pointy-top:
//    /\
//   |  |
//    \/
export class Page {
  radius: number = 0;
  grid: Record<number, Cell> = {};
  p: Coordinates;

  constructor(radius: number, p: Coordinates) {
    this.radius = radius;
    this.p = p;

    for (let x = - this.radius + 1; x < this.radius - 1; x++) {
      for (let y = - this.radius + 1; y < this.radius - 1; y++) {
        this.grid[this.key({ x, y, z: 0-x-y })] = null;
      }
    }
    // console.log('grid size:', Object.keys(this.grid).length);
  }

  static fromObject(page: Page) : Page {
    const newPage = new Page(page.radius, page.p);
    newPage.grid = page.grid;

    return newPage;
  }

  private key(p: Coordinates) : number {
    return p.x + 4 * this.radius * p.y + 16 * this.radius**2 * p.z;
  }

  incByX(p: Coordinates) : Coordinates {
    return {
      x: p.x,
      y: p.y-1,
      z: p.z+1
    };
  }
  decByX(p: Coordinates) : Coordinates {
    return {
      x: p.x,
      y: p.y+1,
      z: p.z-1
    };
  }
  incByY(p: Coordinates) : Coordinates {
    return {
      x: p.x-1, 
      y: p.y,
      z: p.z+1
    };
  }
  decByY(p: Coordinates) : Coordinates {
    return {
      x: p.x+1,
      y: p.y,
      z: p.z-1
    };
  }
  incByZ(p: Coordinates) : Coordinates {
    return {
      x: p.x+1,
      y: p.y-1,
      z: p.z
    };
  }
  decByZ(p: Coordinates) : Coordinates {
    return {
      x: p.x-1,
      y: p.y+1,
      z: p.z
    };
  }

  getNeibhourCoordinates(p: Coordinates) : Coordinates[] {
    return [
      this.incByX(p),
      this.decByX(p),
      this.incByY(p),
      this.decByY(p),
      this.incByZ(p),
      this.decByZ(p)
    ]
    .filter(p => Math.abs(p.x) < this.radius && Math.abs(p.y) < this.radius && Math.abs(p.z) < this.radius);
  }

  getNeibhours(p: Coordinates) : {p: Coordinates, cell: Cell}[] {
    return this.getNeibhourCoordinates(p)
      .map(p2 => ({ p: p2, cell: this.get(p2) }));
  }

  get(p: Coordinates) : Cell {
    return this.grid[this.key(p)];
  }
  
  set(p: Coordinates, value: Cell) : void {
    if (value) {
      this.grid[this.key(p)] = {
        type: value.type,
        playerID: value.playerID,
        component_id: (value && value.component_id !== undefined) ? value.component_id : null,
        p: value.p,
        isBase: (value.isBase !== null || value.isBase !== undefined) ? value.isBase : false
      };
    }
    else {
      this.grid[this.key(p)] = null;
    }
  }

  getRandomEmptyCellCoordinates(): Promise<Coordinates> {
    return new Promise((resolve) => {
      while(true) {
        const p : Coordinates = {
          x : Math.floor(Math.random() * (2 * this.radius - 1)) - this.radius + 1,
          y : Math.floor(Math.random() * (2 * this.radius - 1)) - this.radius + 1,
          z: 0
        };
        p.z = 0 - p.x - p.y;
        if (Math.abs(p.z) >= this.radius) {
          continue;
        }

        const key = this.key(p);
        if (!(key in this.grid) || this.grid[key] === null) {
          resolve(p);
          break;
        }
      }
    });
  }

  getPlayerCells(playerID: string) : Cell[] {
    return Object.values(this.grid).filter(c => c !== null && c !== undefined && c.playerID === playerID);
  }

  removePlayer(playerID: string): void {
    const ps = this.getPlayerCells(playerID).map(c => c.p.cell);
    for (const p of ps) {
      this.set(p, null);
    }
  }
}

File: webbugs-server-go/models/page.go
package models

import (
	"fmt"
	"math"
	"math/rand"
	"sync"

	"github.com/google/uuid"
)

// Page - type for a page
type Page struct {
	Radius uint
	Grid   *sync.Map
	Crd    Coordinates
}

// NewPage - creates new page
func NewPage(radius uint, crd Coordinates) Page {
	var grid sync.Map
	newPage := Page{
		Radius: radius,
		Crd:    crd,
		Grid:   &grid,
	}

	for x := -int64(radius) + 1; x < int64(radius)-1; x++ {
		for y := -int64(radius) + 1; y < int64(radius)-1; y++ {
			newPage.Grid.Store(newPage.key(NewCoordinates(x, y, 0-x-y)), nil)
		}
	}

	return newPage
}

func (page *Page) key(crd Coordinates) int64 {
	return int64(crd.X) +
		4*int64(page.Radius)*int64(crd.Y) +
		16*int64(math.Pow(float64(page.Radius), 2))*int64(crd.Z)
}

// Get - retrieves a cell
func (page *Page) Get(crd Coordinates) *Cell {
	cell, ok := page.Grid.Load(page.key(crd))
	if ok && cell != nil {
		return cell.(*Cell)
	}

	return nil
}

// Set - sets or updates a cell
func (page *Page) Set(crd FullCoordinates, request *CellSetRequest) error {
	if request != nil {
		key := page.key(crd.Cell)
		celli, exists := page.Grid.Load(key)
		if exists && celli != nil {
			cell := celli.(*Cell)
			if crd != cell.Crd {
				return fmt.Errorf("page: set: Unmatching coordinates: %v, %v", crd, cell.Crd)
			}

			cell.FillWithCellSetRequest(*request)
		} else {
			newCell := FromCellSetRequest(*request, crd, page)
			page.Grid.Store(key, &newCell)
		}
	} else {
		key := page.key(crd.Cell)
		page.Grid.Store(key, nil)
	}

	return nil
}

func incByX(crd Coordinates) Coordinates {
	return Coordinates{
		X: crd.X,
		Y: crd.Y - 1,
		Z: crd.Z + 1,
	}
}
func decByX(crd Coordinates) Coordinates {
	return Coordinates{
		X: crd.X,
		Y: crd.Y + 1,
		Z: crd.Z - 1,
	}
}
func incByY(crd Coordinates) Coordinates {
	return Coordinates{
		X: crd.X - 1,
		Y: crd.Y,
		Z: crd.Z + 1,
	}
}
func decByY(crd Coordinates) Coordinates {
	return Coordinates{
		X: crd.X + 1,
		Y: crd.Y,
		Z: crd.Z - 1,
	}
}
func incByZ(crd Coordinates) Coordinates {
	return Coordinates{
		X: crd.X + 1,
		Y: crd.Y - 1,
		Z: crd.Z,
	}
}
func decByZ(crd Coordinates) Coordinates {
	return Coordinates{
		X: crd.X - 1,
		Y: crd.Y + 1,
		Z: crd.Z,
	}
}

// GetNeibhourCoordinates - retrieves neighbour cell coordinates
func (page *Page) GetNeibhourCoordinates(crd Coordinates) []Coordinates {
	ns := [6]Coordinates{
		incByX(crd),
		decByX(crd),
		incByY(crd),
		decByY(crd),
		incByZ(crd),
		decByZ(crd),
	}

	result := make([]Coordinates, 6)
	i := 0

	for _, ncrd := range ns {
		if math.Abs(float64(ncrd.X)) < float64(page.Radius) && math.Abs(float64(ncrd.Y)) < float64(page.Radius) && math.Abs(float64(ncrd.Z)) < float64(page.Radius) {
			result[i] = ncrd
			i = i + 1
		}
	}

	return result
}

func AreNeighbours(crd0 Coordinates, crd1 Coordinates) bool {
	return (crd0.X == crd1.X && math.Abs(float64(crd0.Y)-float64(crd1.Y)) == 1) ||
		(crd0.Y == crd1.Y && math.Abs(float64(crd0.X)-float64(crd1.X)) == 1) ||
		(crd0.Z == crd1.Z && math.Abs(float64(crd0.Y)-float64(crd1.Y)) == 1)
}

// GetNeibhours - retrieves neighbour cells
func (page *Page) GetNeibhours(crd Coordinates) []*Cell {
	crds := page.GetNeibhourCoordinates(crd)

	result := make([]*Cell, len(crds))

	for i, crd2 := range crds {
		result[i] = page.Get(crd2)
	}

	return result
}

// GetOwnWallNeibhours - retrieves neighbour wall cells of a player
func (page *Page) GetOwnWallNeibhours(crd Coordinates, playerID uuid.UUID) []*Cell {
	crds := page.GetNeibhourCoordinates(crd)

	result := make([]*Cell, 0)

	for _, crd2 := range crds {
		cell := page.Get(crd2)
		if cell != nil && cell.PlayerID == playerID && cell.CellType == CellTypeWall {
			result = append(result, cell)
		}
	}

	return result
}

// GetRandomEmptyCellCoordinates - retrieves coordinates of a random empty cell
func (page *Page) GetRandomEmptyCellCoordinates(result chan Coordinates) {
	for {
		p := Coordinates{
			X: int64(rand.Int63n(2*int64(page.Radius)-1)) - int64(page.Radius) + 1,
			Y: int64(rand.Int63n(2*int64(page.Radius)-1)) - int64(page.Radius) + 1,
			Z: 0,
		}
		p.Z = 0 - p.X - p.Y

		if math.Abs(float64(p.Z)) >= float64(page.Radius) {
			continue
		}

		key := page.key(p)
		cell, exists := page.Grid.Load(key)
		if !exists || cell == nil {
			result <- p
			break
		}
	}
}

// GetPlayerCells - retrieves all player cells
func (page *Page) GetPlayerCells(playerID uuid.UUID) []*Cell {
	result := make([]*Cell, 0)
	page.Grid.Range(func(key interface{}, celli interface{}) bool {
		if celli != nil {
			cell := celli.(*Cell)
			if cell.PlayerID == playerID {
				result = append(result, cell)
			}
		}
		return true
	})

	return result
}

// RemovePlayerCells - removes all player cells
func (page *Page) RemovePlayerCells(playerID uuid.UUID) {
	cells := page.GetPlayerCells(playerID)
	for _, cell := range cells {
		page.Set(cell.Crd, nil)
	}
}
